# Lab | Data Cleaning and Formatting

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv


# Challenge 1: Data Cleaning and Formatting

## Exercise 1: Cleaning Column Names

To ensure consistency and ease of use, standardize the column names of the dataframe. Start by taking a first look at the dataframe and identifying any column names that need to be modified. Use appropriate naming conventions and make sure that column names are descriptive and informative.

*Hint*:
- *Column names should be in lower case*
- *White spaces in column names should be replaced by `_`*
- *`st` could be replaced for `state`*

In [43]:
#Import Libraries

import numpy as np
import pandas as pd  

In [44]:
#Read data from file and display the first 5 rows

file1 = pd.read_csv('file1.csv')# Your code here
display (file1.head())

#Shows total rows and columns and makes sure we have a dataframe for file1
print(file1.shape)
type(file1)

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

(4008, 11)


pandas.core.frame.DataFrame

In [45]:
#make names lower case - example
s = "My string"
print("String displayed in lowercase: ",s.lower())
print("Original string: ",s)
s = s.lower()
print("Finally modifed string: ",s)

String displayed in lowercase:  my string
Original string:  My string
Finally modifed string:  my string


In [46]:
#Determine Number of columns
file1.shape[1]

11

In [47]:
range(11)

range(0, 11)

In [48]:
# Make columns uniformly lower case

cols = []
for column in file1.columns:
    cols.append(column.lower())
file1.columns = cols
display (file1.head())

customer          st gender             education customer lifetime value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly premium auto number of open complaints     policy type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle class  total claim amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [49]:
# Remove Spaces
file1.columns = [col.replace(" ", "_") for col in file1.columns]
display (file1.head())

customer          st gender             education customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [50]:
# Change st to state
file1.columns = [col.replace("st", "state") if "st" in col else col for col in file1.columns]




In [51]:
# Change affected customer columns from the previous operation
file1.columns = [col.replace("custateomer", "customer") if "custateomer" in col else col for col in file1.columns]

display (file1.head())

customer       state gender             education customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

## Exercise 2: Cleaning invalid Values

The dataset contains columns with inconsistent and incorrect values that could affect the accuracy of our analysis. Therefore, we need to clean these columns to ensure that they contain only valid data. 

Note that this exercise will focus only on cleaning inconsistent values and will not involve handling null values (NaN or None).

*Hint*:
- *Gender column contains various inconsistent values such as "F", "M", "Femal", "Male", "female", which need to be standardized, for example, to "M" and "F".*
- *State abbreviations can replaced with its full name, for example "AZ": "Arizona", "Cali": "California", "WA": "Washington"*
- *In education, "Bachelors" could be replaced by "Bachelor"*
- *In Customer Lifetime Value, delete the `%` character*
- *In vehicle class, "Sports Car", "Luxury SUV" and "Luxury Car" could be replaced by "Luxury"*

In [52]:

# 1. Cleans the Gender column
gender_new = {
    'Femal': 'F', 
    'Male': 'M', 
    'female': 'F', 
    'F': 'F',   # If it's already 'F', keep it as 'F'
    'M': 'M'    # If it's already 'M', keep it as 'M'
}
file1['gender'] = file1['gender'].map(gender_new).fillna(file1['gender'])

# 2. Cleans the State column
state_new = {
    'AZ': 'Arizona',
    'Cali': 'California',
    'WA': 'Washington',
    'Arizona': 'Arizona',       # If it's already 'Arizona', keep it as 'Arizona'
    'California': 'California', # and so on...
    # Add more states if needed
}
file1['state'] = file1['state'].map(state_new).fillna(file1['state'])

# 3. Cleans the Education column
education_new = {
    'Bachelors': 'Bachelor',
    'Bachelor': 'Bachelor' # If it's already 'Bachelor', keep it as 'Bachelor'
}
file1['education'] = file1['education'].map(education_new).fillna(file1['education'])


# 5. Cleans Vehicle Class column
vehicle_class_new = {
    'Sports Car': 'Luxury', 
    'Luxury SUV': 'Luxury', 
    'Luxury Car': 'Luxury',
    'Luxury': 'Luxury'  # If it's already 'Luxury', keep it as 'Luxury'
}
file1['vehicle_class'] = file1['vehicle_class'].map(vehicle_class_new).fillna(file1['vehicle_class'])

# Display the cleaned dataframe
display(file1.head())


customer       state gender             education customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

## Exercise 3: Formatting data types

The data types of many columns in the dataset appear to be incorrect. This could impact the accuracy of our analysis. To ensure accurate analysis, we need to correct the data types of these columns. Please update the data types of the columns as appropriate. 

It is important to note that this exercise does not involve handling null values (NaN or None).

*Hint*:
- *Customer lifetime value should be numeric*
- *Number of open complaints has a wrong format, look at the diffrent values it takes with unique() and take the middle value, for example 1/5/00 should be 5. Number of open complaints is a string, remember you can use split() to deal with it and take the number you need. Finally, since it should be numeric, cast the column to be in its proper type.*

In [53]:
## Check Data Types fo columns

print(file1.dtypes)
file1.head(25)




customer                      object
state                         object
gender                        object
education                     object
customer_lifetime_value       object
income                       float64
monthly_premium_auto         float64
number_of_open_complaints     object
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object


customer       state gender             education customer_lifetime_value  \
0   RB50392  Washington    NaN                Master                     NaN   
1   QZ44356     Arizona      F              Bachelor              697953.59%   
2   AI49188      Nevada      F              Bachelor             1288743.17%   
3   WW63253  California      M              Bachelor              764586.18%   
4   GA49547  Washington      M  High School or Below              536307.65%   
5   OC83172      Oregon      F              Bachelor              825629.78%   
6   XZ87318      Oregon      F               College              538089.86%   
7   CF85061     Arizona      M                Master              721610.03%   
8   DY87989      Oregon      M              Bachelor             2412750.40%   
9   BQ94931      Oregon      F               College              738817.81%   
10  SX51350  California      M               College              473899.20%   
11  VQ65197  California    NaN               College              819719.71%   
12  DP39365  California    NaN                Master              879879.70%   
13  SJ95423     Arizona    NaN  High School or Below              881901.89%   
14  IL66569  California    NaN               College              538443.17%   
15  BW63560      Oregon    NaN              Bachelor              746313.94%   
16  FV94802      Nevada    NaN  High School or Below              256686.78%   
17  OE15005  California    NaN               College              394524.16%   
18  WC83389      Oregon    NaN               College              571033.31%   
19  FL50705  California    NaN  High School or Below              816261.71%   
20  ZK25313      Oregon    NaN  High School or Below              287205.13%   
21  QK46697  Washington      M              Bachelor              617710.93%   
22  YH23384     Arizona    NaN              Bachelor             2412750.40%   
23  TZ98966      Nevada    NaN              Bachelor              245019.10%   
24  HM55802  California    NaN              Bachelor              239210.79%   

     income  monthly_premium_auto number_of_open_complaints     policy_type  \
0       0.0                1000.0                    1/0/00   Personal Auto   
1       0.0                  94.0                    1/0/00   Personal Auto   
2   48767.0                 108.0                    1/0/00   Personal Auto   
3       0.0                 106.0                    1/0/00  Corporate Auto   
4   36357.0                  68.0                    1/0/00   Personal Auto   
5   62902.0                  69.0                    1/0/00   Personal Auto   
6   55350.0                  67.0                    1/0/00  Corporate Auto   
7       0.0                 101.0                    1/0/00  Corporate Auto   
8   14072.0                  71.0                    1/0/00  Corporate Auto   
9   28812.0                  93.0                    1/0/00    Special Auto   
10      0.0                  67.0                    1/0/00   Personal Auto   
11      0.0                 110.0                    1/0/00   Personal Auto   
12  77026.0                 110.0                    1/2/00  Corporate Auto   
13  99845.0                 110.0                    1/1/00  Corporate Auto   
14  83689.0                  70.0                    1/2/00  Corporate Auto   
15  24599.0                  64.0                    1/1/00  Corporate Auto   
16  25049.0                  67.0                    1/0/00   Personal Auto   
17  28855.0                 101.0                    1/0/00   Personal Auto   
18  51148.0                  72.0                    1/0/00   Personal Auto   
19  66140.0                 101.0                    1/0/00  Corporate Auto   
20  57749.0                  74.0                    1/0/00   Personal Auto   
21  61040.0                  79.0                    1/1/00   Personal Auto   
22  14072.0                  71.0                    1/0/00   Personal Auto   
23      0.0               

In [54]:
# 1. Convert Customer Lifetime Value to numeric (if it's not already)
file1['customer_lifetime_value'] = pd.to_numeric(file1['customer_lifetime_value'], errors='coerce')

# 2. Format the Number of Open Complaints
# First, let's look at the unique values to understand the data
print(file1['number_of_open_complaints'].unique())

# If the values are in the mentioned format, then we can split and extract the middle value
file1['number_of_open_complaints'] = file1['number_of_open_complaints'].str.split('/').str[1].astype(int)

# Display the cleaned dataframe (optional)
display(file1.head())


['1/0/00' '1/2/00' '1/1/00' '1/3/00' '1/5/00' '1/4/00' nan]


customer       state gender             education  customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                      NaN   
1  QZ44356     Arizona      F              Bachelor                      NaN   
2  AI49188      Nevada      F              Bachelor                      NaN   
3  WW63253  California      M              Bachelor                      NaN   
4  GA49547  Washington      M  High School or Below                      NaN   

    income  monthly_premium_auto  number_of_open_complaints     policy_type  \
0      0.0                1000.0                        0.0   Personal Auto   
1      0.0                  94.0                        0.0   Personal Auto   
2  48767.0                 108.0                        0.0   Personal Auto   
3      0.0                 106.0                        0.0  Corporate Auto   
4  36357.0                  68.0                        0.0   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

## Exercise 4: Dealing with Null values

Identify any columns with null or missing values. Identify how many null values each column has. You can use the isnull() function in pandas to find columns with null values.

Decide on a strategy for handling the null values. There are several options, including:

- Drop the rows or columns with null values
- Fill the null values with a specific value (such as the column mean or median for numerical variables and mode for categorical variables)
- Fill the null values with the previous or next value in the column
- Fill the null values based on a more complex algorithm or model (we haven't studied this yet)

Implement your chosen strategy to handle the null values. You can use the fillna() function in pandas to fill null values or dropna() function to drop null values.

Verify that your strategy has successfully handled the null values. You can use the isnull() function again to check if there are still null values in the dataset.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling null values. 

After formatting data types, as a last step, convert all the numeric variables to integers using applymap.

In [13]:
# Check for missing values
file1.isna()

customer  state  gender  education  customer_lifetime_value  income  \
0        False  False    True      False                     True   False   
1        False  False   False      False                     True   False   
2        False  False   False      False                     True   False   
3        False  False   False      False                     True   False   
4        False  False   False      False                     True   False   
...        ...    ...     ...        ...                      ...     ...   
4003      True   True    True       True                     True    True   
4004      True   True    True       True                     True    True   
4005      True   True    True       True                     True    True   
4006      True   True    True       True                     True    True   
4007      True   True    True       True                     True    True   

      monthly_premium_auto  number_of_open_complaints  policy_type  \
0                    False                      False        False   
1                    False                      False        False   
2                    False                      False        False   
3                    False                      False        False   
4                    False                      False        False   
...                    ...                        ...          ...   
4003                  True                       True         True   
4004                  True                       True         True   
4005                  True                       True         True   
4006                  True                       True         True   
4007                  True                       True         True   

      vehicle_class  total_claim_amount  
0             False               False  
1             False               False  
2             False               False  
3             False               False  
4             False               False  
...             ...                 ...  
4003           True                True  
4004           True                True  
4005           True                True  
4006           True                True  
4007           True                True  

[4008 rows x 11 columns]

In [14]:
file1.isna().sum()

customer                     2937
state                        2937
gender                       3054
education                    2937
customer_lifetime_value      4008
income                       2937
monthly_premium_auto         2937
number_of_open_complaints    2937
policy_type                  2937
vehicle_class                2937
total_claim_amount           2937
dtype: int64

Customer Lifetime Value contains only null values and can be dropped
It looks like we need to remove 2937 rows
We can drop rows after row index 1040
Save them to new dataframe in case we need the original again

In [65]:
file2 = file1[file1.index <= 1040]
display(file2.head(1040))

customer       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
1035  BU41599  Washington      F                Master   
1036  TK30357  California      F              Bachelor   
1037  NF31087      Nevada      F  High School or Below   
1038  NH16984  California      F              Bachelor   
1039  OS75493  California      F  High School or Below   

      customer_lifetime_value   income  monthly_premium_auto  \
0                         NaN      0.0                1000.0   
1                         NaN      0.0                  94.0   
2                         NaN  48767.0                 108.0   
3                         NaN      0.0                 106.0   
4                         NaN  36357.0                  68.0   
...                       ...      ...                   ...   
1035                      NaN  62739.0                  70.0   
1036                      NaN  90844.0                 118.0   
1037                      NaN  44685.0                  69.0   
1038                      NaN      0.0                  80.0   
1039                      NaN  42589.0                  98.0   

      number_of_open_complaints     policy_type  vehicle_class  \
0                           0.0   Personal Auto  Four-Door Car   
1                           0.0   Personal Auto  Four-Door Car   
2                           0.0   Personal Auto   Two-Door Car   
3                           0.0  Corporate Auto            SUV   
4                           0.0   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
1035                        0.0   Personal Auto   Two-Door Car   
1036                        0.0   Personal Auto            SUV   
1037                        0.0   Personal Auto  Four-Door Car   
1038                        0.0   Personal Auto  Four-Door Car   
1039                        0.0  Corporate Auto  Four-Door Car   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
1035          239.328571  
1036          232.674417  
1037          331.200000  
1038          336.509610  
1039          470.400000  

[1040 rows x 11 columns]

In [60]:
file1[file1['customer_lifetime_value'].isna()==True].tail(2940) # checking rows that are null based on a specific column

customer       state gender education  customer_lifetime_value   income  \
1068  GS98873     Arizona      F  Bachelor                      NaN  16061.0   
1069  CW49887  California      F    Master                      NaN  79487.0   
1070  MY31220  California      F   College                      NaN  54230.0   
1071      NaN         NaN    NaN       NaN                      NaN      NaN   
1072      NaN         NaN    NaN       NaN                      NaN      NaN   
...       ...         ...    ...       ...                      ...      ...   
4003      NaN         NaN    NaN       NaN                      NaN      NaN   
4004      NaN         NaN    NaN       NaN                      NaN      NaN   
4005      NaN         NaN    NaN       NaN                      NaN      NaN   
4006      NaN         NaN    NaN       NaN                      NaN      NaN   
4007      NaN         NaN    NaN       NaN                      NaN      NaN   

      monthly_premium_auto  number_of_open_complaints    policy_type  \
1068                  88.0                        0.0  Personal Auto   
1069                 114.0                        0.0   Special Auto   
1070                 112.0                        0.0  Personal Auto   
1071                   NaN                        NaN            NaN   
1072                   NaN                        NaN            NaN   
...                    ...                        ...            ...   
4003                   NaN                        NaN            NaN   
4004                   NaN                        NaN            NaN   
4005                   NaN                        NaN            NaN   
4006                   NaN                        NaN            NaN   
4007                   NaN                        NaN            NaN   

      vehicle_class  total_claim_amount  
1068  Four-Door Car               633.6  
1069            SUV               547.2  
1070   Two-Door Car               537.6  
1071            NaN                 NaN  
1072            NaN                 NaN  
...             ...                 ...  
4003            NaN                 NaN  
4004            NaN                 NaN  
4005            NaN                 NaN  
4006            NaN                 NaN  
4007            NaN                 NaN  

[2940 rows x 11 columns]

In [67]:
print(file2.isnull().sum())

customer                        0
state                           0
gender                        117
education                       0
customer_lifetime_value      1041
income                          0
monthly_premium_auto            0
number_of_open_complaints       0
policy_type                     0
vehicle_class                   0
total_claim_amount              0
dtype: int64


There's still a problem with gender. A reasonable solution might be to fill in this categorical 50/50 M/F as that is broadly reflective of the distribution in the population.

In [68]:
# Find the indices of the rows with null values in the 'gender' column
null_indices = file2[file2['gender'].isnull()].index

# Calculate the number of null values
num_nulls = len(null_indices)

# Split that number in half
half_num_nulls = num_nulls // 2

# Assign half of the null values to 'M' and the other half to 'F'
file2.loc[null_indices[:half_num_nulls], 'gender'] = 'M'
file2.loc[null_indices[half_num_nulls:], 'gender'] = 'F'


In [69]:
# Check if gender now has no null values
print(file2.isnull().sum())

customer                        0
state                           0
gender                          0
education                       0
customer_lifetime_value      1041
income                          0
monthly_premium_auto            0
number_of_open_complaints       0
policy_type                     0
vehicle_class                   0
total_claim_amount              0
dtype: int64


In [71]:
display(file2)

customer       state gender             education  \
0     RB50392  Washington      M                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
1036  TK30357  California      F              Bachelor   
1037  NF31087      Nevada      F  High School or Below   
1038  NH16984  California      F              Bachelor   
1039  OS75493  California      F  High School or Below   
1040  VT63298      Oregon      M              Bachelor   

      customer_lifetime_value   income  monthly_premium_auto  \
0                         NaN      0.0                1000.0   
1                         NaN      0.0                  94.0   
2                         NaN  48767.0                 108.0   
3                         NaN      0.0                 106.0   
4                         NaN  36357.0                  68.0   
...                       ...      ...                   ...   
1036                      NaN  90844.0                 118.0   
1037                      NaN  44685.0                  69.0   
1038                      NaN      0.0                  80.0   
1039                      NaN  42589.0                  98.0   
1040                      NaN      0.0                  72.0   

      number_of_open_complaints     policy_type  vehicle_class  \
0                           0.0   Personal Auto  Four-Door Car   
1                           0.0   Personal Auto  Four-Door Car   
2                           0.0   Personal Auto   Two-Door Car   
3                           0.0  Corporate Auto            SUV   
4                           0.0   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
1036                        0.0   Personal Auto            SUV   
1037                        0.0   Personal Auto  Four-Door Car   
1038                        0.0   Personal Auto  Four-Door Car   
1039                        0.0  Corporate Auto  Four-Door Car   
1040                        0.0   Personal Auto  Four-Door Car   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
1036          232.674417  
1037          331.200000  
1038          336.509610  
1039          470.400000  
1040          345.600000  

[1041 rows x 11 columns]

## Exercise 5: Dealing with duplicates

Use the .duplicated() method to identify any duplicate rows in the dataframe.

Decide on a strategy for handling the duplicates. Options include:
- Dropping all duplicate rows
- Keeping only the first occurrence of each duplicated row
- Keeping only the last occurrence of each duplicated row
- Dropping duplicates based on a subset of columns
- Dropping duplicates based on a specific column

Implement your chosen strategy using the drop_duplicates() function.

Verify that your strategy has successfully handled the duplicates by checking for duplicates again using .duplicated().

Remember to document your process and explain your reasoning for choosing a particular strategy for handling duplicates.

Save the cleaned dataset to a new CSV file.

*Hint*: *after dropping duplicates, reset the index to ensure consistency*.

In [73]:
# Show duplicated rows
duplicate_rows = file2[file2.duplicated()]
display(duplicate_rows)


Empty DataFrame
Columns: [customer, state, gender, education, customer_lifetime_value, income, monthly_premium_auto, number_of_open_complaints, policy_type, vehicle_class, total_claim_amount]
Index: []

In [74]:
# Drop rows that are duplicated in customer, other rows can have valid duplicates
file2.drop_duplicates(subset='customer')

customer       state gender             education  \
0     RB50392  Washington      M                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
1036  TK30357  California      F              Bachelor   
1037  NF31087      Nevada      F  High School or Below   
1038  NH16984  California      F              Bachelor   
1039  OS75493  California      F  High School or Below   
1040  VT63298      Oregon      M              Bachelor   

      customer_lifetime_value   income  monthly_premium_auto  \
0                         NaN      0.0                1000.0   
1                         NaN      0.0                  94.0   
2                         NaN  48767.0                 108.0   
3                         NaN      0.0                 106.0   
4                         NaN  36357.0                  68.0   
...                       ...      ...                   ...   
1036                      NaN  90844.0                 118.0   
1037                      NaN  44685.0                  69.0   
1038                      NaN      0.0                  80.0   
1039                      NaN  42589.0                  98.0   
1040                      NaN      0.0                  72.0   

      number_of_open_complaints     policy_type  vehicle_class  \
0                           0.0   Personal Auto  Four-Door Car   
1                           0.0   Personal Auto  Four-Door Car   
2                           0.0   Personal Auto   Two-Door Car   
3                           0.0  Corporate Auto            SUV   
4                           0.0   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
1036                        0.0   Personal Auto            SUV   
1037                        0.0   Personal Auto  Four-Door Car   
1038                        0.0   Personal Auto  Four-Door Car   
1039                        0.0  Corporate Auto  Four-Door Car   
1040                        0.0   Personal Auto  Four-Door Car   

      total_claim_amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
1036          232.674417  
1037          331.200000  
1038          336.509610  
1039          470.400000  
1040          345.600000  

[1041 rows x 11 columns]

# Bonus: Challenge 2: creating functions on a separate `py` file

Put all the data cleaning and formatting steps into functions, and create a main function that performs all the cleaning and formatting. 

Write these functions in a separate .py file(s). By putting these steps into functions, we can make the code more modular and easier to maintain. 

*Hint: autoreload module is a utility module in Python that allows you to automatically reload modules in the current session when changes are made to the source code. This can be useful in situations where you are actively developing code and want to see the effects of changes you make without having to constantly restart the Python interpreter or Jupyter Notebook kernel.*

In [ ]:
# Your code here

# Bonus: Challenge 3: Analyzing Clean and Formated Data

You have been tasked with analyzing the data to identify potential areas for improving customer retention and profitability. Your goal is to identify customers with a high policy claim amount and a low customer lifetime value.

In Pandas Lab, we only looked at high policy claim amounts because we couldn't look into low customer lifetime values. If we tried to do work with that column, we wouldn't have been able to do it because customer lifetime value wasn't clean and in its proper format. So after cleaning and formatting the data, lets get some more interesting insights!

Instructions:

- Review again the statistics for total claim amount and customer lifetime value to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount and a low customer lifetime value. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount, and clients with a low customer lifetime value to be those in the bottom 25% of the customer lifetime value. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile and a customer lifetime value in the bottom 25th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount and low customer lifetime value data. To do so, select both columns of the dataframe simultaneously and pass it to the .describe() method. This will give you descriptive statistics, such as mean, median, standard deviation, minimum and maximum values for both columns at the same time, allowing you to compare and analyze their characteristics.

In [ ]:
# Your code here